### PR TITLE
Ryan Budd: Project05 Vertex Shader

### PR DESCRIPTION
**Pull request by RBudd16005**
Setup the Matrix class and its helper functions. Defined the transformationVertexShader. Everything works fine except for the last triangle, I am not quite sure why.
screenshot 10

### CODE REVIEW

**Comment by sta14031** [comment]
Your STR function is not correctly performing the multiplication of a 4x4 matrix times a 4x4 matrix to combine all the transformations that it needs. This is why the fourth triangle comes out distorted.

File: coursefunctions.h [hunk 1]
```diff
@@ -429,42 +429,59 @@ void TestVertexShader(Buffer2D<PIXEL> & target)
...
+        colorUniforms.other[1] = 100;
+        colorUniforms.other[2] = 50;
 
 	DrawPrimitive(TRIANGLE, target, colorTriangle, colorAttributes, &colorUniforms, &myColorFragShader, &myColorVertexShader);
 
         /***********************************
          * SCALE (scale by a factor of 0.5)
          ***********************************/
-        // Your scaling code that integrates with 'colorUniforms', used by 'myColorVertexShader' goes here
+        colorUniforms.other[0] = 1;
```
Comment (Justin9four) [inline coursefunctions.h]: leave comment here to show that other[0] is your "what transformation am I doing" attribute

File: definitions.h [hunk 2]
```diff
@@ -282,18 +267,238 @@ class Attributes
...
+        //Scale the vertex by the uniform using matrix multiplication
+        void scale(const Vertex & vertIn, const Attributes & uniforms)
+        {
+            double sum[4] = {0, 0, 0, 0};
+            double vert[4] = {vertIn.x, vertIn.y, vertIn.z, vertIn.w};
+
+            //Setup the matrix
+            four[0][0] = uniforms.other[1];
+            four[1][1] = uniforms.other[1];
+
```
Comment (Justin9four) [inline definitions.h]: it would be cool to override the multiplication operator as we discussed in class,  to be more human readable

**Comment by Bluhm-Alexander** [comment]
The reason why your last triangle won't work is because you need to do to matrix math operations and your only doing one matrix math operation.

For example:
colorUniforms.transforms = {
                {cos(45 * (M_PI/180)), -(sin(45 *(M_PI/180))), 0, 0},
                {(sin(45 *(M_PI/180))), cos(45 * (M_PI/180)), 0, 0},
                {0, 0, 1, 0},
                {0, 0, 0, 1}
        };

        Matrix scale;
        scale = {
                {.5, 0, 0, 100},
                {0, .5, 0, 50},
                {0, 0, 1, 0},
                {0, 0, 0, 1}
        };

//Then you multiply the two matrices together

colorUniforms.transforms = colorUniforms.transforms * scale;

You do your rotation first then multiply is by scale and then translation is last.

**Comment by RBudd16005** [comment]
Fixed it! I was forgetting to use a new matrix every multiplication, so I was accidentally performing every transformation twice.

**Pull request merged**